core/blockchain.ex

```diff
@@ -23,8 +26,40 @@ defmodule UltraDark.Blockchain do
 
     Ledger.append_block(block)
     UtxoStore.update_with_transactions(block.transactions)
+	rebalance_difficulty? chain
 
     chain
   end
 
+  def rebalance_difficulty(chain) do
+	last = List.first(chain)
+	beginning = Enum.at(chain, max(length(chain) - @diff_rebalance_offset + 1, 0))
+	avg_spb = (unix_timestamp(last.timestamp) - unix_timestamp(beginning.timestamp)) / @diff_rebalance_offset
+	speed_ratio = @target_blocktime / avg_spb
+	prev = last.difficulty
+
+	# difficulty = log speed_ratio base 16 = log2(speed_ratio) / log2(16)
+	diff = :math.log2(speed_ratio) / 4
+	chain = List.update_at(chain, 0, &(%{&1 | difficulty: diff}))
```

Comment:
youre changing the difficulty of an already mined block here, thatll make the block hash invalid

Reply:
@alexdovzhanyn which block _should_ I be setting the difficulty of?

Reply:
this should be setting the difficulty variable to another value, and blocks that are mined in the future will read from that value. essentially, create a new variable that's stored in the `main` function of the miner that stores the current difficulty

Reply:
ah okay. might it be worth making the blockchain into a struct?

Reply:
i agree, although it would warrant a rewrite of some of the ways we're passing the chain around

Reply:
I'll have a look at doing it, but if it's too much work I might leave it for another PR